### PR TITLE
[SYCL][CUDA] Fix LIT fails on machines without non-NVIDIA OpenCL

### PR DESCRIPTION
**Pull request by bjoernknafla**
We are observing LIT fails on machines that only have the NVIDIA OpenCL, even if running for the SYCL PI CUDA backend.

These commits try to fix the problem but still require testing.

### CODE REVIEW

File: sycl/test/Unit/lit.cfg.py [hunk 1]
```diff
@@ -66,3 +66,7 @@ def find_shlibpath_var():
 else:
     lit_config.warning("unable to inject shared library path on '{}'"
                        .format(platform.system()))
+
+config.environment['SYCL_BE'] = lit_config.params.get('SYCL_BE', "PI_OPENCL")
+
+lit_config.note("Environment: {}".format(config.environment))
```
Comment (bjoernknafla) [inline sycl/test/Unit/lit.cfg.py:72]: I wonder if there is a better way to print the implicit environment with which LIT runs the unit tests, e.g., one that allows a copy and paste approach to setting up the environment locally to investiage fails?

File: sycl/test/basic_tests/get_nonhost_devices.cpp [hunk 2]
```diff
@@ -1,6 +1,8 @@
-// RUN: %clangxx -fsycl %s -o %t.out
-// RUN: %t.out
-
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/get_nonhost_devices.cpp:2]: using SYCL_DEVICE_TYPE affects only default_selector (see https://github.com/intel/llvm/blob/sycl/sycl/source/device_selector.cpp#L89)

There is no reason for the change.

File: sycl/test/Unit/lit.cfg.py (hunk 1 above)
Comment (vladimirlaz) [reply sycl/test/Unit/lit.cfg.py:72]: That looks like debug info. I would prefer to remove it from final sumission.

Comment (bjoernknafla) [reply sycl/test/Unit/lit.cfg.py:72]: I am not too happy about this either. I am using it in ` sycl/test/basic_tests/diagnostics/device-check.cpp` (see changes just below). 

Alternative could be to only run the test below for OpenCL (by adding `REQUIRES: opencl`) as it tests a backend independent code path anyway.

What do you prefer?

File: sycl/test/basic_tests/get_nonhost_devices.cpp (hunk 2 above)
Comment (bjoernknafla) [reply sycl/test/basic_tests/get_nonhost_devices.cpp:2]: On a machine with just the NVIDIA OpenCL installed the above fails (sadly on an automatic testing machine I do not have debugging access to).

The `XYZ_RUN_PLACEHOLDER` also adds the `SYCL_BE=PI_XYZ` environment variable and replaces the CPU and ACC runs. With just an NVIDIA OpenCL installed a `check-sycl-cuda` test run then avoids OpenCL and passes.

Alternatively, I could mark the test as `REQUIRES: opencl` so it would be ignored in `check-sycl-cuda` test runs.

Comment (vladimirlaz) [reply sycl/test/basic_tests/get_nonhost_devices.cpp:2]: "REQUIRES: opencl" is more clear from my perspective.

Comment (bjoernknafla) [reply sycl/test/basic_tests/get_nonhost_devices.cpp:2]: Thank you, I will go with that.

File: sycl/test/Unit/lit.cfg.py (hunk 1 above)
Comment (bjoernknafla) [reply sycl/test/Unit/lit.cfg.py:72]: I am not awake - completely missed what you tried to say here 🤦 

You are right, this is information to enable debugging of a failed test as otherwise the environment in which the unit tests are run is invisible which makes replication of bugs very hard.

File: sycl/test/basic_tests/get_nonhost_devices.cpp (hunk 2 above)
Comment (bjoernknafla) [reply sycl/test/basic_tests/get_nonhost_devices.cpp:2]: Done. Will be in next push.

File: sycl/test/Unit/lit.cfg.py (hunk 1 above)
Comment (bjoernknafla) [reply sycl/test/Unit/lit.cfg.py:72]: I will replace this with output of the `SYCL_BE` environment as we also do in the non-unit tests: [sycl/test/lit.cfg.py#L76](https://github.com/intel/llvm/blob/sycl/sycl/test/lit.cfg.py#L76)

Comment (bjoernknafla) [reply sycl/test/Unit/lit.cfg.py:72]: The user then still gets the required info to re-run unittests by hand to recreate fails.

Comment (vladimirlaz) [reply sycl/test/Unit/lit.cfg.py]: Could you clarify which info is missed? if you look into failing tests  (e.g. http://ci.llvm.intel.com:8010/#/builders/37/builds/689/steps/15/logs/FAIL__SYCL__reduction_nd_conditional_cpp):
 - 'RUN: at line 4';    env SYCL_DEVICE_TYPE=GPU SYCL_BE=PI_CUDA  /localdisk2/sycl_ci/buildbot/worker/Lit_With_Cuda/llvm.obj/tools/sycl/test/reduction/Output/reduction_nd_conditional.cpp.tmp.out
You see the command line with specific environment variables set.

**Comment by bjoernknafla** [comment]
The reduction fails of [buildbot/Lit_With_Cuda](http://ci.llvm.intel.com:8010/#/builders/37/builds/696)  have been fixed and merged 2h ago: https://github.com/intel/llvm/pull/1641

File: sycl/test/Unit/lit.cfg.py (hunk 1 above)
Comment (bjoernknafla) [reply sycl/test/Unit/lit.cfg.py]: There are two approaches of setting environment variables in use:

* explicitly (as shown by your example) as can bee seen in the lit.cfg.py file here: [sycl/test/lit.cfg.py#L127](https://github.com/intel/llvm/blob/sycl/sycl/test/lit.cfg.py#L127) - these will show up when a LIT test fails, and
* implicitly by manipulating the environment a LIT test will run in as seen here: [sycl/test/lit.cfg.py#L45](https://github.com/intel/llvm/blob/sycl/sycl/test/lit.cfg.py#L45) - these are not shown when a LIT test fails.

Comment (bjoernknafla) [reply sycl/test/Unit/lit.cfg.py]: My current thinking about this is:
* Use the implicit environment variables to pass the existing environment through. Failing tests will run in the same environment (ignoring CI/buildbot systems for now) when running them by hand. Otherwise LIT filters out most environment variables: [llvm/utils/lit/lit/TestingConfig.py#L24](https://github.com/intel/llvm/blob/sycl/llvm/utils/lit/lit/TestingConfig.py#L24)

* Use the explicit environment variable approach for settings that are very specific to the test and test setup and that we configure inside of LIT as these will be visible when a test fails (or when running lit with the `-a` flag) and allow copy-and-paste rerunning.

Sadly the explicit approach does not work when running unit tests as there is no way foreseen by LIT (that I know of) to set the environment explicitly when running GTest tests...

File: sycl/test/basic_tests/get_nonhost_devices.cpp [hunk 3]
```diff
@@ -1,6 +1,9 @@
-// RUN: %clangxx -fsycl %s -o %t.out
+// REQUIRES: opencl
```
Comment (vladimirlaz) [inline sycl/test/basic_tests/get_nonhost_devices.cpp:1]: We can add cuda here,

Comment (bjoernknafla) [reply sycl/test/basic_tests/get_nonhost_devices.cpp:1]: We have seen this test fail on machines that do not have OpenCL drivers but only CUDA installed. One fix might be to add `env SYCL_BE=%sycl_be` to the run line but we need to test this to see if it works. Adding the `%GPU_RUN_PLACEHOLDER` fixed the test, too.

Comment (bjoernknafla) [reply sycl/test/basic_tests/get_nonhost_devices.cpp]: Changed.

File: sycl/test/basic_tests/diagnostics/handler.cpp [hunk 4]
```diff
@@ -1,5 +1,8 @@
-// RUN: %clangxx -fsycl %s -o %t.out
-// RUN: %t.out | FileCheck %s
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t.out | FileCheck %s
+// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
+// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
+// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```
Comment (bader) [inline sycl/test/basic_tests/diagnostics/handler.cpp:5]: Are these changes needed?
https://github.com/intel/llvm/pull/1543 changed only line 1. Isn't it enough?

Is it related to the issue in sycl/test/basic_tests/get_nonhost_devices.cpp? If so, should it be resolved the same way?

Comment (bjoernknafla) [reply sycl/test/basic_tests/diagnostics/handler.cpp:5]: My understanding is that the test code triggers the default selector implicitly when it creates the `Queue`.  Therefore the `PLACEHOLDER` substitutions are used (for `RUN` and `CHECK`) to more explicitly control what is tested.

File: sycl/test/basic_tests/queue.cpp [hunk 5]
```diff
@@ -1,6 +1,8 @@
 // RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
 // RUN: env SYCL_DEVICE_TYPE=HOST %t.out
-// RUN: %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
```
Comment (bader) [inline sycl/test/basic_tests/queue.cpp:5]:
```suggestion
// RUN: env SYCL_BE=%sycl_be %t.out
```

File: sycl/test/basic_tests/diagnostics/handler.cpp (hunk 4 above)
Comment (bader) [reply sycl/test/basic_tests/diagnostics/handler.cpp:5]:
```suggestion
// RUN: env SYCL_BE=%sycl_be %t.out | FileCheck %s
```
Should we really run this test on 4 different devices to validate exceptions handling?

I'd like to keep existing testing approach and align on overall testing strategy separately.

Comment (vladimirlaz) [reply sycl/test/basic_tests/diagnostics/handler.cpp:5]: +1 The check and exception generation is done on host side (before offloading). There is no reason to run it on every device.

Comment (bjoernknafla) [reply sycl/test/basic_tests/diagnostics/handler.cpp:5]: I see your point. Change with next pull.

File: sycl/test/basic_tests/queue.cpp (hunk 5 above)
Comment (bjoernknafla) [reply sycl/test/basic_tests/queue.cpp]: Fixed.
